src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java

```diff
@@ -48,38 +50,74 @@ public WithHoldingOrderLineUpdateInstanceStrategy(InventoryItemManager inventory
-          // TODO: onSuccess or compose ???
-          .onSuccess(v -> deleteAbandonedHoldings(replaceInstanceRef.getDeleteAbandonedHoldings(), holder.getStoragePoLine(), requestContext));
-        case CREATE -> createHoldingsAndUpdateItems(holder, newInstanceId, requestContext)
-          .onSuccess(v -> deleteAbandonedHoldings(replaceInstanceRef.getDeleteAbandonedHoldings(), holder.getStoragePoLine(), requestContext));
-        default -> Future.succeededFuture();
-      };
-    } else {
+    ReplaceInstanceRef replaceInstanceRef = holder.getPatchOrderLineRequest().getReplaceInstanceRef();
+    if (replaceInstanceRef == null) {
       return Future.succeededFuture();
     }
+    String newInstanceId = replaceInstanceRef.getNewInstanceId();
+
+    holder.createStoragePatchOrderLineRequest(StoragePatchOrderLineRequest.PatchOrderLineOperationType.REPLACE_INSTANCE_REF, newInstanceId);
+    PoLine poLine = holder.getStoragePoLine();
+
+    return switch (replaceInstanceRef.getHoldingsOperation()) {
+      case MOVE -> moveHoldings(poLine, newInstanceId, requestContext);
+      case FIND_OR_CREATE -> findOrCreateHoldingsAndUpdateItems(holder, newInstanceId, requestContext)
+        .compose(v -> deleteAbandonedHoldings(replaceInstanceRef.getDeleteAbandonedHoldings(), poLine, requestContext));
+      case CREATE -> createHoldingsAndUpdateItems(holder, newInstanceId, requestContext)
+        .compose(v -> deleteAbandonedHoldings(replaceInstanceRef.getDeleteAbandonedHoldings(), poLine, requestContext));
+      default -> Future.succeededFuture();
+    };
   }
 
 
-  private Future<Void> moveHoldings(OrderLineUpdateInstanceHolder holder, String newInstanceId, RequestContext requestContext) {
-    return pieceStorageService.getPiecesByPoLineId(PoLineCommonUtil.convertToCompositePoLine(holder.getStoragePoLine()), requestContext)
-      .map(pieces -> extractUniqueHoldingIds(pieces, holder.getStoragePoLine().getLocations()))
-      .compose(holdingIds -> {
-        holdingIds.forEach(holdingId -> holder.addHoldingRefsToStoragePatchOrderLineRequest(holdingId, holdingId));
-        return inventoryHoldingManager.getHoldingsByIds(holdingIds, requestContext);
-      })
-      .compose(holdings -> {
-        removeHoldingUnrecognizedFields(holdings);
-        return inventoryHoldingManager.updateInstanceForHoldingRecords(holdings, newInstanceId, requestContext);
-      });
+  private Future<Void> moveHoldings(PoLine poLine, String newInstanceId, RequestContext requestContext) {
+    CompositePoLine compPOL = PoLineCommonUtil.convertToCompositePoLine(poLine);
+    return pieceStorageService.getPiecesByPoLineId(compPOL, requestContext)
+      .map(pieces -> getHoldingsByTenants(poLine, pieces, requestContext))
+      .compose(holdingsByTenant -> {
+          var updateHoldings = holdingsByTenant.entrySet()
+            .stream()
+            .map(entry -> entry.getValue().compose(holdings -> {
+              removeHoldingUnrecognizedFields(holdings);
```

Comment:
this logic with multitenancy is complicated, you can look through changes to add more logs that can help us in troubleshooting

Reply:
Good point, working on this.